platforms/documentation/docs/src/docs/userguide/optimizing-builds/configuration_cache.adoc

```diff
@@ -536,6 +536,57 @@ And, when an incompatible task is scheduled to run, Gradle discards the configur
 
 Check the method documentation for more details.
 
+[[config_cache:integrity_check]]
+=== Using integrity checks to debug store and load issues
+By default, the configuration cache doesn't perform many integrity checks when reading the entry.
+This improves performance and reduces entry size.
+However, it makes it harder to pinpoint some store and load issues.
+For example, the incorrectly stored object may be discovered later, when reading data of some unrelated task.
+It may cause a type mismatch when restoring a value of the property or break some other invariant, resulting in a potentially misleading exception.
+
+In order to make debugging easier, it is possible to enable stricter checks that may help to discover discrepancies earlier.
+To do that, add the following line to your `gradle.properties`:
+[source,properties]
+.gradle.properties
+----
+org.gradle.configuration-cache.integrity-check=true
+----
+
+For example, let's look at a type that implements a custom serialization protocol incorrectly:
+[source,java]
+----
+include::{snippetsPath}/configurationCache/integrityChecks/common/buildSrc/src/main/java/User.java[tags=broken-type]
+----
+<1> `writeObject` serializes both fields.
+<2> `readObject` reads only the first field, leaving the lingering data.
+
+Such a type will cause problems when used as part of a task state.
+[source,java]
+----
+include::{snippetsPath}/configurationCache/integrityChecks/common/buildSrc/src/main/java/GreetTask.java[tags=task-with-broken-type]
+----
+
+When running without the integrity check, you may get a cryptic failure message, perhaps accompanied by some induced configuration cache problems:
```

Comment:
🤔 Now that we have a flag, it makes me wonder if we should suggest using it in such cases. Even if it's a heuristic for allow-listed set of exceptions, it would already point users in the right direction of the integrity checks. Otherwise, they are expected to know or read through the manual, and we all know what this usually means.

Reply:
> Now that we have a flag, it makes me wonder if we should suggest using it in such cases.

We can do it as part of adopting the Problems API in CC.